### PR TITLE
[System]: The big WebConnectionStream cleanup.

### PR DESCRIPTION
**Pull request by baulig**
`WebResponseStream` has been broken down into smaller pieces and now
fully supports GZip / Deflate decompression.

The entire `contentLength`, `readBuffer` and `totalRead` logic has been
replaced by a nested layer of `WebReadStream` classes.

* `WebReadStream` is a new abstract base class.

* `BufferedReadStream` handles the situation where we have some buffered
  data and an (optional) inner stream to read from.

* `ContentDecodeStream` takes care of GZip / Deflate decompression.

* `FixedSizeReadStream` handles the `contentLength` logic - ie. making sure
  we are only reading up to `contentLength` bytes from the stream.

* The old `MonoChunkStream` has been renamed into `MonoChunkParser` (as it was
  really just a parser and not a stream) and a new `MonoChunkStream` has been
  added that's based on `WebReadStream`.

### CODE REVIEW

**Comment by baulig** [comment]
@monojenkins commit csproj

**Comment by luhenry** [comment]
@marek-safar could you please review?

File: mcs/class/System/System.Net/WebReadStream.cs [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+	{
+		public WebOperation Operation {
+			get;
+		}
+
+		protected Stream InnerStream {
+			get;
+		}
+
+		protected readonly string ME;
```
Comment (marek-safar) [inline mcs/class/System/System.Net/WebReadStream.cs]: You should make that a method as it's just debug helper

File: mcs/class/System/System.Net/WebReadStream.cs [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+		{
+			Operation = operation;
+			InnerStream = innerStream;
+
+#if MONO_WEB_DEBUG
+			ME = $"WRS({GetType ().Name}:Op={operation.ID})";
+#endif
+		}
+
+		Exception NotSupported => new NotSupportedException ();
```
Comment (marek-safar) [inline mcs/class/System/System.Net/WebReadStream.cs]: What's the benefit of this extra method?

File: mcs/class/System/System.Net/WebOperation.cs [hunk 3]
```diff
@@ -63,7 +63,7 @@ class WebOperation
 		string ME => $"WO({ID},{Connection?.ID ?? -1})";
 #else
 		internal readonly int ID;
-		string ME = null;
+		readonly string ME = null;
```
Comment (marek-safar) [inline mcs/class/System/System.Net/WebOperation.cs]: This should be method as well

File: mcs/class/System/System.Net/WebReadStream.cs [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+		{
+			throw NotSupported;
+		}
+
+		public override void Flush ()
+		{
+			throw NotSupported;
+		}
+
+		protected Exception GetException (Exception e)
```
Comment (marek-safar) [inline mcs/class/System/System.Net/WebReadStream.cs]: should be static

**Comment by baulig** [comment]
Creating new, clean PR.

**Comment by baulig** [comment]
Oops, wrong PR.

**Comment by baulig** [comment]
The close / reopen was an accident, sorry about that.  @marek-safar I addressed your issues, could you please comment again?

**Comment by marek-safar** [comment]
@baulig does it also resolve https://bugzilla.xamarin.com/show_bug.cgi?id=59779 ?

**Comment by baulig** [comment]
Yes, could you please close that one for me, I somehow have problems logging into bugzilla?

**Pull request merged**